util/ctxkey/key.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// ctxkey provides type-safe key-value pairs for use with [context.Context].
+//
+// Example usage:
+//
+//	// Create a context key.
+//	var DeleteCooldownKey = ctxkey.New("logz.DeleteCooldown", 5*time.Second)
+//
+//	// Store a context value.
+//	ctx = logz.DeleteCooldownKey.WithValue(ctx, 10*time.Second)
+//
+//	// Load a context value.
+//	cooldown := logz.DeleteCooldownKey.Value(ctx)
+//	... // use cooldown of type time.Duration
+//
+// This is inspired by https://go.dev/issue/49189.
+package ctxkey
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+)
+
+// Key is a generic key type associated with a specific value type.
+// The Value type must not be an interface type.
+type Key[Value any] struct {
+	name   *string
+	defVal *Value
+}
+
+// New constructs a new context key with an associated value type
+// where the default value for an unpopulated value is the provided value.
+//
+// The provided name is an arbitrary name only used for human debugging.
+// Every key is unique, even if provided the same name.
+// As a convention, it is recommended that the name be the dot-delimited
+// combination of the package name of the caller with the variable name.
+//
+// Example usage:
+//
+//	package mapreduce
+//	var NumWorkersKey = ctxkey.New("mapreduce.NumWorkersKey", runtime.NumCPU())
+func New[Value any](name string, value Value) Key[Value] {
```

Comment:
maybe rename second parameter to `defaultValue` to make it's use clear (in addition to what's in the godoc)?

Reply:
Done.